### ISSUE TITLE
Document causes frozen terminal, 100% CPU

### ISSUE THREAD

**Issue by ijl** [jwilm/alacritty#2326]
If I run `curl https://raw.githubusercontent.com/miloyip/nativejson-benchmark/master/data/canada.json`, the terminal becomes unresponsive and consumes 100% of CPU. It may at first be responsive, stalling after ~10s, or with -vvv becomes unresponsive sometimes before it finishes rendering the file.

Memory usage is unchanged. It seems to consume CPU indefinitely. 

The file is a ~2MiB JSON document of mostly floats. There are no formatting characters in it.

This is on alacritty 0.3.0 and current HEAD (ab8fddd5938024beca3a97eab9022db07f675a3c) on Linux and macOS with default configuration.

With alacritty -vvv here are two different logs:

```
...
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1330] Scrolling up relative: origin=0, lines=1
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1694] Linefeed
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1330] Scrolling up relative: origin=0, lines=1
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: Bold
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: Reverse
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelReverse
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: Bold
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: Reset
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: Reset
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelReverse
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelUnderline
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1921] Clearing screen: Below
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1865] Clearing line: Right
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2045] Setting mode: BracketedPaste
[2019-04-18 22:25] [DEBUG] Resizing terminal
[2019-04-18 22:25] [DEBUG] New num_cols is 479 and num_lines is 65
[2019-04-18 22:25] [INFO] Width: 3840, Height: 1055
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: Reset
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelReverse
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelUnderline
[2019-04-18 22:25] [TRACE] [src/term/mod.rs:1921] Clearing screen: Below
[end]
```

```
...
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1330] Scrolling up relative: origin=0, lines=1
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1694] Linefeed
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1330] Scrolling up relative: origin=0, lines=1
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: Bold
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: Reverse
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelReverse
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: Bold
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: Reset
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1686] Carriage return
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: Reset
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelReverse
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2012] Setting attribute: CancelUnderline
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1921] Clearing screen: Below
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:1865] Clearing line: Right
[2019-04-18 22:29] [TRACE] [src/term/mod.rs:2045] Setting mode: BracketedPaste
[end]
```

**Comment by chrisduerr** [comment]
Which shell are you using? For me on Linux/X11 with zsh this is no problem at all.

**Comment by ijl** [comment]
Hm, I'm using zsh 5.7.1.

**Comment by felipesere** [comment]
I'm also on zsh 5.7.1, using Alacritty 0.3.0 from homebrew and it also hangs after curling that file.

**Comment by chrisduerr** [comment]
Sounds like this is a macOS specific issue. Can someone confirm that this does not happen in iTerm2?

**Comment by ijl** [comment]
The logs above are from Linux. I reported the issue after seeing it on both Linux and macOS.

**Comment by chrisduerr** [comment]
This might be a duplicate of https://github.com/jwilm/alacritty/issues/2217, though I'm not sure. Could someone confirm this doesn't happen in a different terminal emulator?

Also when running `alacritty -vvv`, does it keep printing things even after it looks like it froze?

**Comment by felipesere** [comment]
I just tested it again with iTerm2 3.2.1beta3 and Terminal.app 2.9.1 and they work fine.

I have attached a screen recording of Alacritty hanging after downloading that JSON file.
Sadly, it doesn't show the spinning-beachball :( 
When I circle the mouse, I was seeing the spinning beachball that signals that a Mac app has hung or is "not responding".

![download-json-bug](https://user-images.githubusercontent.com/1850188/56467551-4f10bb00-6418-11e9-94f4-e3b0039ced3c.gif)

**Comment by chrisduerr** [comment]
Oh huh, I thought this was actually about Alacritty not showing that things are downloading at all. This is kinda different.

Can you type `reset` in the terminal and then press enter while it's frozen? Even though it doesn't show, it might still execute and unfreeze Alacritty.

**Comment by felipesere** [comment]
I tried `reset` and nothing happend. It doesn't seem to register any keystrokes anymore.
Is there a way to get more output than `alacritty -vvvv`?

**Comment by chrisduerr** [comment]
Using `alacritty -vvv --print-events` will also print all events as they are received by Alacritty. So that might be helpful to troubleshoot if it's receiving anything at all.

If this is a deadlock, like the spinning symbol would indicate, it might be interesting to check with a debugger where it's hanging.

**Comment by ijl** [comment]
This appears resolved on 0.3.3.

**Comment by anonymous4argus** [comment]
It seems the issue still exists in 0.3.3 but get fixed in 0.4.0. Base on tracing for the execution, it is the same issue as #3002, which is caused by the excessive memory operations in UrlParser.
